### PR TITLE
Use pkg-config variables

### PR DESCRIPTION
**Pull request by emersion**
This allows all variables to be set properly when the prefix or datadir changes.

The generated .pc file looks like this:

    prefix=/usr/local
    datadir=${prefix}/share
    completionsdir=${datadir}/fish/vendor_completions.d
    functionsdir=${datadir}/fish/vendor_functions.d
    confdir=${datadir}/fish/vendor_conf.d

    Name: fish
    Description: fish, the friendly interactive shell
    URL: https://fishshell.com/
    Version: 3.1.0-402-g75ae172ba228-dirty

Closes: https://bugs.archlinux.org/task/65904

### CODE REVIEW

**Comment by zanchey** [comment]
We discussed this extensively in #5029; the `pkg-config` directories are for installations where the package manager is not involved (eg installation from source).

It's still possible we haven't gotten this right though, there's a lot of gnarliness. If you haven't read through that and the related issues, would you be able to do so and revise your PR if required?

**Comment by emersion** [comment]
>the pkg-config directories are for installations where the package manager is not involved

Other software relies on `pkg-config` to find the fish completions dir: https://github.com/swaywm/sway/blob/master/meson.build#L281

I don't know why one would want to install `pkg-config` files with erroneous paths in it.

File: cmake/Install.cmake [hunk 1]
```diff
@@ -24,17 +24,21 @@ set(configure_input
...
  DO NOT MANUALLY EDIT THIS FILE!")
 
+set(extra_datadir
+    "\${prefix}/${rel_datadir}"
+    CACHE STRING "pkg-config path for datadir")
+
 set(extra_completionsdir
-    /usr/local/share/fish/vendor_completions.d
-    CACHE STRING "Path for extra completions")
+    "\${datadir}/fish/vendor_completions.d"
```
Comment (faho) [inline cmake/Install.cmake]: This will result in

```fish
set __extra_completionsdir ${datadir}/fish/vendor_completions.d
```

in [share/__fish_build_paths.fish](https://github.com/fish-shell/fish-shell/blob/master/share/__fish_build_paths.fish.in),  which is invalid fish code. For fish, you'd want to use `\$datadir` here (similar for $prefix), which results in

```fish
set __extra_completionsdir $datadir/fish/vendor_completions.d
```

I am not sure if pkg-config can deal with that form. If not, it would have to be adjusted.

Comment (faho) [reply cmake/Install.cmake]: Thinking further about it, keeping these variables unexpanded will *not* work for __fish_build_paths at all, because they won't be defined when that's executed.

So we would have to compute them for that. I assume computing them would also work for pkg-config?

Comment (emersion) [reply cmake/Install.cmake]: Ah, I'll fix that.

>So we would have to compute them for that. I assume computing them would also work for pkg-config?

No, because if the user installs e.g. Sway in a different prefix, the user doesn't want Sway to try to install anything (completions) outside of this prefix. See https://github.com/swaywm/swaybg/issues/13

Comment (faho) [reply cmake/Install.cmake]: Wait. What?

So you install bash with a prefix of /usr. It computes a datadir of /usr/share, and a completionsdir of /usr/share/bash-completion/completions.

Now you install sway with a prefix of /usr/local. Do you expect the sway bash completions to go to /usr/local/share/bash-completion/completions, and bash to pick that up? How is bash to know? If it knows /usr/local, how is it supposed to know /banana/?

I feel like I'm missing something quite major here - all of my experience building software is the happy path of making packages to use system-wide, so I just use /usr for everything.

So I would expect "the directory where bash completions go" to be a system-wide thing (plus a different "the directory where your own completions go")

Comment (emersion) [reply cmake/Install.cmake]: > Do you expect the sway bash completions to go to /usr/local/share/bash-completion/completions

Yes. No files outside of the prefix.

> and bash to pick that up? How is bash to know? If it knows /usr/local, how is it supposed to know /banana/?

It's up to the user to configure the system-wide bash copy to pick stuff in their custom prefix. It's the same as `LD_LIBRARY_PATH`, `PKG_CONFIG_PATH` and friends needed when installing a library to a custom prefix.

pkg-config variables like `${prefix}` get automatically replaced with the custom prefix by Meson.

**Comment by krobelus** [comment]
Hi,

TLDR: I think I agree that the paths provided by package config should be within the prefix by default, but I think I read that the packaging systems Homebrew and Nix need them outside. So pkg-config might be flexible enough on its own.

I found this topic *really* tricky. It looks like the current situation is still unsatisfactory.

In any case, fish will source completions from both paths (from `/usr` as well as from `/usr/local`), unless you override that in `$XDG_DATA_DIRS`

I can see your point that the pkg-config path should be below the prefix for sane packaging.
The default for `extra_completionsdir` is a bit nicer for software that is being installed from source and drops fish completions there.

It is a question of who we optimize for. With the current situation for software like `grim` that ships fish completions, the distribution that packages it has to work around this by moving the completions from /usr/local to /usr when creating the package. Which sucks because packaging should be easy to get right.

I *think* that it would make more sense for the Arch package to set `extra_completionsdir` to `/usr/share/fish/vendor_completions.d`, since Arch does not use /usr/local at all by default.
If that happens then the distribution's job is easier, but installing some software with `make PREFIX=/usr/local install` might install fish completions to /usr/share which doesn't seem reasonable. With this example, it'd seem logical to install fish completions to `$PREFIX/share/fish/vendor_completions.d` instead of /usr/share/. That might solve both issues, because distribution packages use PREFIX=/usr and installations from source use PREFIX=/usr/local or similar.

That would require hardcoding `share/fish/vendor_completions.d` which I see [you already have](https://github.com/emersion/grim/blob/master/contrib/completions/meson.build), but only as fallback. What you do seems to be reasonable:

1. If there is a fish installation in `$PREFIX`, use its pkg-config value.
2. Otherwise fall back to `$PREFIX/share/fish/vendor_completions.d`.

*However*, in the above scenario pkg-config may use the fish installed to a /usr, even when installing to /usr/local. This is a feature, but it doesn't seem desirable in this case. So as I understand it *not using pkg-config*, but only the fallback would do the "right thing".

The original motivation for making these paths configurable was #2113.
For those use cases it probably is essential that other software is using the pkg-config variables.

Other packagers, like you want pkg-config to provide a value that is below the prefix where fish is installed. Sounds like conflicting use cases.

I'd find it attractive to recommend to *always* install vendor completions to `$PREFIX/share/fish/vendor_completions.d` (fully independent of where fish is installed), since the pkg-config solution can't really cover all use cases. This means we'd shift the problem to somewhere else, probably the user that needs to update their completion path when they use a nonstandard prefix, which seems okay.

File: cmake/Install.cmake (hunk 1 above)
Comment (krobelus) [reply cmake/Install.cmake]: Yeah it would be nice to have a simpler model here, but I don't see know how that would work with pkg-config, because pkg-config could use a fish.pc from a different prefix than the one you're installing to?

**Comment by krobelus** [comment]
@zanchey stated this problem already in https://github.com/fish-shell/fish-shell/pull/6508#issuecomment-576257385

> The problem is there's actually no easy right answer here for developers of other packages - the best we can do is say "use pkg-config, but allow it to be overridden by (eg) dpkg-buildpackage/rpmbuild", as when built by package managers they should install to the prefix, but outside of that should generally go in /usr/local .

So for packages, like `grim` the current status means that
1. developers need to provide a configuration option to override pkg-config
2. distributions need to use that configuration option to install fish files to prefix

This sounds really bad because it can affect many packages in future.

I don't think it's a big problem that source installs drop files in `/usr/share/fish/vendor_completions.d`. Or is it?
When someone installs `grim` both using their package manager, and from source, they need to resolve the conflict manually and decide for either of the two completion files,. This seems reasonable to me.

So I feel like this should be accepted because it can make packaging much simpler at what seems like low cost.

File: cmake/Install.cmake (hunk 1 above)
Comment (krobelus) [reply cmake/Install.cmake]: Just to illustrate: so if you install fish to prefix `/fish` and grim to prefix `/grim` with your patch then `grim.fish` will be installed to `/fish/share/fish/vendor_completions.d/` (assuming you told pkg-config about the fish.pc). This seems fine. Currently it would be `/usr/local/share/fish/vendor_completions.d`.

**Comment by zanchey** [comment]
This effectively reverts 81306d3b300c732a70a15ba5e5a90a817cbfc5fa and reopens #5029.

**Comment by zanchey** [comment]
(I think? Sorry)

**Comment by krobelus** [comment]
You mean that `/usr/local/share/fish/vendor_completions.d` should be in `$fish_complete_path`?
I think it will still be there because we now add `${XDG_DATA_DIRS:-/usr/local/share:/usr/share}/fish/vendor_completions.d`.
So I don't see how this reintroduces #5029 unless you mean something else.

**Comment by emersion** [comment]
Okay, updated the pull request to be less intrusive and only change the pkg-config file, not the `extra_*` variables. I don't think the "extra" prefix should be referenced at all from the pkg-config file.

**Comment by zanchey** [comment]
I finally understand this PR! Sorry for taking so long.

So, I *think* this is probably right, or as close to right as we can get. I do wonder a bit about the wisdom of unconditionally including `/usr/local` in the paths, but it seems like a reasonable thing to do and I guess we can see how it pans out.

The CHANGELOG entry probably needs fleshing out, to something like "The pkg-config file now uses pkg-config variables to point to prefix-specific directories; if you install third-party packages in a location other than fish's prefix or `/usr/local`, you may need to adjust your `fish_{complete,function}_paths."

Additionally, there is no way of setting fish to look in extra paths for `conf.d` configuration snippets, so that's going to need addressing if we do merge this.

**Comment by emersion** [comment]
>I do wonder a bit about the wisdom of unconditionally including /usr/local in the paths, but it seems like a reasonable thing to do and I guess we can see how it pans out.

Prior to this PR, `/usr/local` was hardcoded in all paths. With this PR, it's not hardcoded in the pkg-config file anymore.

>Additionally, there is no way of setting fish to look in extra paths for conf.d configuration snippets, so that's going to need addressing if we do merge this.

I don't think this PR changes anything about this. It only touches the pkg-config file.

**Comment by krobelus** [comment]
> I do wonder a bit about the wisdom of unconditionally including /usr/local in the paths

That feels a bit weird for systems that use `$XDG_DATA_DIRS` because `/usr/local` can't be excluded by simply dropping it from that variable. For other system it's probably a good idea, overall it's fine I think.

File: fish.pc.in [hunk 2]
```diff
@@ -1,8 +1,8 @@
 prefix=@prefix@
-datadir=@datadir@
-completionsdir=@extra_completionsdir@
-functionsdir=@extra_functionsdir@
-confdir=@extra_confdir@
+datadir=${prefix}/@rel_datadir@
+completionsdir=${datadir}/@rel_completionsdir@
```
Comment (faho) [inline fish.pc.in:3]: So, this will just not work because, while the completionsdir will now be set in the pkg-config file, and so the completions will be *installed* there, fish won't *read* them because __fish_build_paths will still contain the paths in /usr/local.

Comment (emersion) [reply fish.pc.in:3]: This patch doesn't change the completions installation path, nor the paths where fish searches completions.

Comment (faho) [reply fish.pc.in:3]: Yes, but it changes where things would be installed via pkg-config.

Comment (faho) [reply fish.pc.in:3]: Without altering where fish *looks* - that's what `__fish_build_paths.fish` contains.

Comment (emersion) [reply fish.pc.in:3]: But fish already searches in `/usr/share/fish/vendor_completions.d`, right? `extra_completionsdir` is just an _extra_ directory.

Comment (faho) [reply fish.pc.in:3]: No, vendor_completions.d *is* extra_completionsdir.

Comment (emersion) [reply fish.pc.in:3]: Before this patch:

* System completions are installed in `/usr/share/fish/vendor_completions.d`
* They are looked up by fish
* `/usr/local/share/fish/vendor_completions.d` is used in the pkg-config file but doesn't exist
* `extra_completionsdir` is set to `/usr/local/share/fish/vendor_completions.d`

```
> ls /usr/share/fish/vendor_completions.d
alacritty.fish  curl.fish  hub.fish  pass.fish  rg.fish  sway.fish  swayidle.fish  swaylock.fish  swaymsg.fish  swaynag.fish  yay.fish  youtube-dl.fish
> ls /usr/local/share/fish/vendor_completions.d
ls: cannot access '/usr/local/share/fish/vendor_completions.d': No such file or directory
> cat /usr/share/pkgconfig/fish.pc 
…
completionsdir=/usr/local/share/fish/vendor_completions.d
…
```

Comment (faho) [reply fish.pc.in:3]: Why are the completions installed in /usr/share/fish/vendor_completions.d if the pkg-config file says /usr/local/share/fish/vendor_completions.d?

The reason this also looks up /usr/share/fish/vendor_completions.d is because of $XDG_DATA_DIRS containing /usr/share, so we add that to the searchpath.

But what *should* be changed is that extra_completionsdir should typically point to /usr/share/fish/vendor_completions.d, with "/usr" being the prefix.

Currently, it will change the pkg-config entry but not the fish variable, so if the prefix is set to something that *isn't* in $XDG_DATA_DIRS the directory won't be searched.

Comment (emersion) [reply fish.pc.in:3]: >Why are the completions installed in /usr/share/fish/vendor_completions.d if the pkg-config file says /usr/local/share/fish/vendor_completions.d?

Because currently no-one uses pkg-config to find where completions should be installed. And it's a good thing since doing so would install to `/usr/local`, which is definitely not something a distro package should do.

>The reason this also looks up /usr/share/fish/vendor_completions.d is because of $XDG_DATA_DIRS containing /usr/share, so we add that to the searchpath.

Oh, I see. Okay, let's change that too. People will still be able to override it via build options anyway.

**Comment by emersion** [comment]
Done and rebased.

**Comment by faho** [comment]
Yeah, this looks good to me.

Merged, thanks!
